File: TwoPointers/167. Two Sum II - Input array is sorted/twoSumII.js
/**
 * @param {number[]} numbers
 * @param {number} target
 * @return {number[]}
 */
// map
var twoSum = function (numbers, target) {
    let map = new Map();
    for (let i = 0; i < numbers.length; i++) {
        map.set(numbers[i], i + 1);
    }

    for (let i = 0; i < numbers.length; i++) {
        let diff = target - numbers[i];
        if (map.has(diff) && map.get(diff) !== i + 1) {
            return [i + 1, map.get(diff)]
        }
    }
};



/**
 * @param {number[]} numbers
 * @param {number} target
 * @return {number[]}
 */
// Two pointers
var twoSum = function (numbers, target) {
    let left = 0;
    let right = numbers.length - 1;

    while (left < right) {
        if (numbers[left] + numbers[right] > target) {
            right--;
        }

        if (numbers[left] + numbers[right] < target) {
            left++;
        }

        if (numbers[left] + numbers[right] === target) {
            break;
        }
    }

    return [left + 1, right + 1];
};

File: TwoPointers/167. Two Sum II - Input array is sorted/twoSumII.py
# map
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        dict = {}
        for i, num in enumerate(numbers):
            dict[numbers[i]] = i + 1
        
        for i, num in enumerate(numbers):
            diff = target - num
            if diff in dict and dict[diff] != i:
                return [i + 1, dict[diff]]

# two pointers
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        left = 0
        right = len(numbers) - 1
        
        while left < right:
            if numbers[left] + numbers[right] > target:
                right -= 1
            if numbers[left] + numbers[right] < target:
                left += 1
            if numbers[left] + numbers[right] == target:
                break
        
        return [left + 1, right + 1]